### ISSUE TITLE
Delay js change in 3.9 breaks elementor mobile menus

### ISSUE THREAD

**Issue by jazir555** [wp-media/wp-rocket#3931]
**Describe the bug**
The new delay js breaks elementor mobile menu toggles for the dropdown entirely. I've excluded every single js file in elementor and they still dont work.

**To Reproduce**
1. enable delay js
2. Place a mobile menu element in a header in elementor theme builder
3. check the menu and see that its broken

**Expected behavior**
Elementors mobile menu should not break. I can't even fix it by excluding every js file that elementor has.

**Backlog Grooming (for WP Media dev team use only)**
- [ ] Reproduce the problem
- [ ] Identify the root cause
- [ ] Scope a solution
- [ ] Estimate the effort

**Comment by GeekPress** [comment]
@jazir5 Can you please share a URL of a website using elementor menu? Thanks :)

**Comment by githuphub** [comment]
I'm facing similar issues on mobile, everything from Elementor is excluded but the top banner image and slider are not showing until I scroll.

**Comment by GeekPress** [comment]
> everything from Elementor is excluded but the top banner image and slider are not showing until I scroll.

@githuphub This is the expected behavior and the concept of Delay JS; JS files are loaded after user interaction (scroll, click, etc...).

**Comment by jazir555** [comment]
@GeekPress 

I would like to request that you re-add the option to manually specify files to be delayed, while also keeping the current option as a separate checkmark to delay all scripts. 

Excluding all scripts has also broken some woocommerce functionality like add to cart, some products add 2 to cart when the button has been clicked once, then adds one product to cart after the initial 2 products.

>everything from Elementor is excluded but the top banner image and slider are not showing until I scroll.

I believe he's saying these are above the fold elements that are broken until user interaction

**Comment by GeekPress** [comment]
@jazir5 Can you please share a URL with elementor menu? It will be very helpful, thanks in advance :)

**Comment by jazir555** [comment]
@GeekPress 

Sure will do, as soon as the designer finishes work on the site for the day. Don't want to break it, I reverted to 3.8.8 for the time being.

**Comment by jazir555** [comment]
@GeekPress 

I've updated wp-rocket and you can go to my site and see the broken mobile menu now:

https://www.lonecbd.com

It also breaks the header dropdown on desktop, shop is supposed to have a dropdown

My mini-cart widget is also completely broken

Remove unused css is not active

**Comment by jazir555** [comment]
If you could check soon that would be great, I'd like to revert asap so my site isn't broken until the next update :)

**Comment by jazir555** [comment]
@Tabrisrp

**Comment by engahmeds3ed** [comment]
I got a screenshot of the issue and sent it to @GeekPress internally so u can revert that as we started the weekend holiday.

I can see that you have wpmeteor plugin enabled, what if you disabled it and check again?

**Comment by jazir555** [comment]
> I got a screenshot of the issue and sent it to @GeekPress internally so u can revert that as we started the weekend holiday.
> 
> I can see that you have wpmeteor plugin enabled, what if you disabled it and check again?

Awesome thank you. Yes I do have wp-meteor active, this also seems to be an incompatibility with wp-meteor. Is there anyway you could update wp-rocket to be compatible with it? To clarify, disabling wp-meteor *does* seem to fix it, but only on desktop.

**Comment by jazir555** [comment]
Also, before I disable it, could you check my discounts page and also the about page? The pages break temporarily then work fine

**Comment by engahmeds3ed** [comment]
I can see them working fine now, can you please contact us at our support email so we can solve those issues with you?
I think you disabled wpmeteor now, I can see the whole site working properly, right?

**Comment by jazir555** [comment]
Not quite, if you load the pages on mobile you can see them visually break. In addition, the mobile menu seems to be working, but requires two taps to open as the delay causes it to not work without user interaction.

Any chance you could update wp-rocket to be compatible with wp-meteor? It really helps the pagespeed on product pages

**Comment by jazir555** [comment]
Sure, what's the email address?

**Comment by jazir555** [comment]
It's actually the same with the mini cart as well, takes two taps to open it

**Comment by engahmeds3ed** [comment]
support[at]wp-rocket.me
or from here
https://wp-rocket.me/support/

**Comment by jazir555** [comment]
Do you you get/see the same issues with the other pages on mobile? Same with the double taps being required for the menu and mini-cart? Those are kind of critical to fix to be able to use the new delay feature for me

**Comment by jazir555** [comment]
I'm going to revert in a couple hours or so, please do let me know if you can reproduce the issues I mentioned.

**Comment by engahmeds3ed** [comment]
Yes, I can see that happens if u didn't interact with the page (scroll for example).
Please contact us to help you, don't worry I got your point.

**Comment by jazir555** [comment]
Great, thanks. Do you also see the page breaking issues on the other pages, just to confirm? It only breaks on mobile. 

Already emailed you :)

**Comment by engahmeds3ed** [comment]
I saw it one time but I can't see it now.
We will check your email ASAP.

**Comment by jazir555** [comment]
I appreciate you being so responsive and helping me troubleshoot, thank you.

**Comment by jazir555** [comment]
Oh just to clarify, it seems to happen on every page load if you load it in
incognito modeq

**Comment by jazir555** [comment]
Every first page load in incognito*

**Comment by gayanchandrab** [comment]
Excluding JS will not work because this new Delay JS script delays DOMContentLoaded event. I'm having trouble with Slider Revolution. Even if all related JavaScript code is excluded, slider will not load because DOMContentLoaded is delayed until user interaction.

**Comment by jazir555** [comment]
> Excluding JS will not work because this new Delay JS script delays DOMContentLoaded event. I'm having trouble with Slider Revolution. Even if all related JavaScript code is excluded, slider will not load because DOMContentLoaded is delayed until user interaction.

This is why I really wish they had both options for delaying js. The old opt-in only method where you specify scripts you want delayed, and the option to enable the new way where it delays all scripts.

**Comment by gayanchandrab** [comment]
Yes. I also wish old Delay JS option can be kept as an option. Also when DOMContentLoaded is delayed, BeTheme's menu will not be visible until user interaction because it waits until DOMContentLoaded is fired to remove a CSS class with display: none rule.

**Comment by jazir555** [comment]
There are so many issues with the new method, I probably have to stay on a lower version until all these are patched. I will re-update to demonstrate the bugs to any wp-rocket team members if needed, but it messes with my site in too many ways to leave active right now

**Comment by mehedihasanziku** [comment]
I don't think double tap issue related to new feature, because in elementor menu js isn't independent, it relays on elementor core js. so if all elementor js excluded from delay menu should work on first tap. anyways it can be fixed through extra custom js, this is an expected behaviour of elementor if all js in delay. i don't think they will provide solution except excluding all elementor JS.

Moreover, i am following this ticket, i think you shouldn't update/downgrade the plugin, create a duplicate/staging site of live, then in staging you can test this feature and support can also give proper look on it.

Thanks.

**Comment by jazir555** [comment]
The double tap issue only appears in 3.9. What files in Elementor should be
excluded to have the double tap not be necessary? If I exclude all
elementor files, the performance is considerably worse.

**Comment by mehedihasanziku** [comment]
yes, because before delay was only ketword based,  not delaying all. if you delay all in old one by putting "elementor" keyword,  you will face exact same issue like new one. 

keyword should be excluded.

frontend
menu
waypoint
webpack
preloaded


of course it will be lower performance. we have to exclude many files of elementor to make it work, as elementor menu not depends on single js file, you can convert your menu to mega menu, which has two js files, it can be easily excluded, and menu will work fine without issues.

**Comment by jazir555** [comment]
Does excluding them currently work, or does it have the same issues as as
the other DOM loaded events and the inline scripts not being able to be
excluded currently?

**Comment by mehedihasanziku** [comment]
Please try & share. i am not listed in 10% unfortunately.

**Comment by jazir555** [comment]
Currently those files must be excluded, if wp-meteor is updated that may
not be the case

**Comment by jazir555** [comment]
I can confirm it is still broken with the files excluded

**Comment by jazir555** [comment]
Even with wp-meteor disabled, the menu doesn't seem to work

**Comment by mehedihasanziku** [comment]
well, then i need to wait to get latest one. i think more js needed to be excluded.

**Comment by GeekPress** [comment]
@jazir5 WP Rocket is 100% compatible with WP Meteor. You have just need to disable Delay JS option... You can't use two systems to delay your JS files. It's like you want to use two caching plugins at the same time, it's not possible.

**Comment by GeekPress** [comment]
@jazir5 Can you please contact me at jonathan @ wp-media.me with WordPress access to your website. I will also need to have the plugin editor available. 

We have a new version of our Delay JS script we would like to test on your site.

**Comment by jazir555** [comment]
Yes, I sent you an email with the login credentials

**Comment by GeekPress** [comment]
@jazir5 I didn't receive an email with website credentials. Please write an email and don't send an automatic WordPress email

**Comment by jazir555** [comment]
I did, I'm surprised you didn't receive it. Might have went to spam. I'll
get back to you as soon as the designer finishes up on our site, dont want
to complicate his work by telling him to stop and then start again

**Comment by Mai-Saad** [comment]
Note: 
The same problem occurs when using Astra theme but excluding `/wp-content/themes/astra/assets/js/minified/frontend.min.js?ver=3.4.7` , fixes the problem.

**Comment by GeekPress** [comment]
@Mai-Saad Did you used the last Dealy JS version available on the repo? If no, please try with the lastest version.

**Comment by Mai-Saad** [comment]
@GeekPress No I was testing on 3.9 zip.. let me check the latest delay js version.

**Comment by Mai-Saad** [comment]
@GeekPress Still not working with the new delay js.

The scenario is as follow while having Astra active theme on mobile view:
1- Disable delay js then clear cache, 
2- Access home page => have cached version (good menu)
3- Enable delay js  => cached version of the home page is auto-updated 
4- Hard refresh home page => Menu is broken ( if we clear cache here then hard refresh page, the menu will be good)

Notes:
- Same will happen to the rest of the pages (using the same steps) if preload is enabled 
- If before step 1 we excluded `/wp-content/themes/astra/assets/js/minified/frontend.min.js?ver=3.4.7` => menu will be good using the same steps
- Working fine on 3.8.8
-  Video: 

https://user-images.githubusercontent.com/76941962/120221675-18d0d700-c23f-11eb-8a20-d476abce69b8.mp4

**Comment by jazir555** [comment]
I can confirm I am using Astra as well

**Comment by piotrbak** [comment]
In my case toggling the mobile Astra menu is a problem. I've investigated this problem and when this file is delayed:
`/wp-content/themes/astra/assets/js/minified/frontend.min.js`

The `resize` event is not registred in the browser:
![image](https://user-images.githubusercontent.com/26767037/120239544-6dcc1780-c256-11eb-8ebb-bd14d56a4a1d.png)

**Comment by GeekPress** [comment]
@piotrbak @Mai-Saad Can we please move the Avada conversation to the right issue?  https://github.com/wp-media/wp-rocket/issues/3941 

It's going to be quite hard to follow if we talk about several page builder/themes in the same issue.

**Comment by GeekPress** [comment]
@jazir5 After visiting your website (https://www.lonecbd.com) I can notice you are using WP Meteor AND our Delay JS. Please note it's not possible to use both. You have to select the solution you prefer.

Regarding the original issue, please update to 3.9.0.2 and let me know if the problem has been fixed.

**Comment by jazir555** [comment]
@GeekPress 

Could you please update my account for wp-rocket(techtutormm@gmail.com) to be able to download 3.9.0.2? For some reason I cant download it

**Comment by GeekPress** [comment]
> @GeekPress
> 
> Could you please update my account for wp-rocket([techtutormm@gmail.com](mailto:techtutormm@gmail.com)) to be able to download 3.9.0.2? For some reason I cant download it

Sure, it's done.

**Comment by kaeelo5** [comment]
hi i am using astra theme and this exact issue is happening to me too. Can someone help what the solution is to the problem?

currently on 3.10.1

**Comment by piotrbak** [comment]
Hello @kaeelo5 

Could you please check if 3.10.2 version resolves your problem? If not, please create a support ticket and we'll investigate this issue further.

**Comment by piotrbak** [comment]
The issue is not there anymore.

**Comment by jasperf** [comment]
We seem to see mobile menu briefly for https://lean-warrior.com/ before the main desktop menu loads with JS deferred and JS execution delayed. So it seems to be a similar issue to the one here. I excluded Astra Theme for execution delay. Does not seem to help as of yet. Excluding it from deferred loading 

```
/wp-content/themes/astra/assets/js/minified/frontend.min.js
```

does not help either. The menu still loads mobile version before loading desktop version on desktop.


![Screenshot 2023-07-24 at 15 16 35](https://github.com/wp-media/wp-rocket/assets/344138/891d0589-62bc-4633-b58f-acb4958b9db3)
